## REGRAS DE ASSOCIAÇÃO EM PYTHON

In [2]:
# Apriori

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


In [3]:
# Data Preprocessing
dataset = pd.read_csv('Market_Basket_Optimisation.csv') 


In [4]:
dataset

shrimp            almonds            avocado  \
0               burgers          meatballs               eggs   
1               chutney                NaN                NaN   
2                turkey            avocado                NaN   
3         mineral water               milk         energy bar   
4        low fat yogurt                NaN                NaN   
5     whole wheat pasta       french fries                NaN   
6                  soup        light cream            shallot   
7     frozen vegetables          spaghetti          green tea   
8          french fries                NaN                NaN   
9                  eggs           pet food                NaN   
10              cookies                NaN                NaN   
11               turkey            burgers      mineral water   
12            spaghetti          champagne            cookies   
13        mineral water             salmon                NaN   
14        mineral water                NaN                NaN   
15               shrimp          chocolate            chicken   
16               turkey               eggs                NaN   
17               turkey         fresh tuna           tomatoes   
18            meatballs               milk              honey   
19             red wine             shrimp              pasta   
20                 rice    sparkling water                NaN   
21            spaghetti      mineral water                ham   
22              burgers      grated cheese             shrimp   
23                 eggs                NaN                NaN   
24      parmesan cheese          spaghetti               soup   
25          ground beef          spaghetti      mineral water   
26      sparkling water                NaN                NaN   
27        mineral water               eggs            chicken   
28    frozen vegetables          spaghetti               yams   
29        herb & pepper       tomato sauce        light cream   
...                 ...                ...                ...   
7470      mineral water          chocolate                NaN   
7471          chocolate      grated cheese      mineral water   
7472   whole wheat rice               cake     cottage cheese   
7473             turkey             shrimp              pasta   
7474             turkey      mineral water        cooking oil   
7475           tomatoes       tomato sauce          spaghetti   
7476       tomato juice                NaN                NaN   
7477             turkey        ground beef      mineral water   
7478             shrimp              pasta      mineral water   
7479  frozen vegetables        french wine           pancakes   
7480          chocolate             shrimp              pasta   
7481          chocolate                NaN                NaN   
7482         fresh tuna    parmesan cheese  whole wheat pasta   
7483            cookies       energy drink                NaN   
7484    parmesan cheese          champagne         white wine   
7485           red wine       tomato sauce          spaghetti   
7486               soup               milk                NaN   
7487               eggs   whole wheat rice                NaN   
7488           brownies                NaN                NaN   
7489      herb & pepper          spaghetti     low fat yogurt   
7490      herb & pepper                NaN                NaN   
7491          chocolate           escalope                NaN   
7492            burgers             salmon           pancakes   
7493             turkey            burgers       dessert wine   
7494           pancakes         light mayo                NaN   
7495             butter         light mayo        fresh bread   
7496            burgers  frozen vegetables               eggs   
7497            chicken                NaN                NaN   
7498           escalope          green tea                NaN   
7499               eggs    frozen smoothie        yo

In [5]:
# preciso tirar os cabeçalhos
# Data Preprocessing
dataset = pd.read_csv('Market_Basket_Optimisation.csv', header = None)


In [6]:
dataset

0                  1                  2   \
0                shrimp            almonds            avocado   
1               burgers          meatballs               eggs   
2               chutney                NaN                NaN   
3                turkey            avocado                NaN   
4         mineral water               milk         energy bar   
5        low fat yogurt                NaN                NaN   
6     whole wheat pasta       french fries                NaN   
7                  soup        light cream            shallot   
8     frozen vegetables          spaghetti          green tea   
9          french fries                NaN                NaN   
10                 eggs           pet food                NaN   
11              cookies                NaN                NaN   
12               turkey            burgers      mineral water   
13            spaghetti          champagne            cookies   
14        mineral water             salmon                NaN   
15        mineral water                NaN                NaN   
16               shrimp          chocolate            chicken   
17               turkey               eggs                NaN   
18               turkey         fresh tuna           tomatoes   
19            meatballs               milk              honey   
20             red wine             shrimp              pasta   
21                 rice    sparkling water                NaN   
22            spaghetti      mineral water                ham   
23              burgers      grated cheese             shrimp   
24                 eggs                NaN                NaN   
25      parmesan cheese          spaghetti               soup   
26          ground beef          spaghetti      mineral water   
27      sparkling water                NaN                NaN   
28        mineral water               eggs            chicken   
29    frozen vegetables          spaghetti               yams   
...                 ...                ...                ...   
7471      mineral water          chocolate                NaN   
7472          chocolate      grated cheese      mineral water   
7473   whole wheat rice               cake     cottage cheese   
7474             turkey             shrimp              pasta   
7475             turkey      mineral water        cooking oil   
7476           tomatoes       tomato sauce          spaghetti   
7477       tomato juice                NaN                NaN   
7478             turkey        ground beef      mineral water   
7479             shrimp              pasta      mineral water   
7480  frozen vegetables        french wine           pancakes   
7481          chocolate             shrimp              pasta   
7482          chocolate                NaN                NaN   
7483         fresh tuna    parmesan cheese  whole wheat pasta   
7484            cookies       energy drink                NaN   
7485    parmesan cheese          champagne         white wine   
7486           red wine       tomato sauce          spaghetti   
7487               soup               milk                NaN   
7488               eggs   whole wheat rice                NaN   
7489           brownies                NaN                NaN   
7490      herb & pepper          spaghetti     low fat yogurt   
7491      herb & pepper                NaN                NaN   
7492          chocolate           escalope                NaN   
7493            burgers             salmon           pancakes   
7494             turkey            burgers       dessert wine   
7495           pancakes         light mayo                NaN   
7496             butter         light mayo        fresh bread   
7497            burgers  frozen vegetables               eggs   
7498            chicken                NaN                NaN   
7499           escalope          green tea                NaN   
7500               eggs    frozen smoothie        yogurt

In [7]:
#transformar os dados em uma lista de listas
transactions = []
for i in range(0, 7501):
    transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])


#### parâmetros 

min_support - indica o mínimo de vezes que devem aparecer nos dados; no caso ele quer que aparecem no mínimo 3 vezes por semana

In [10]:
3*7/7500

0.0028

min_confiance indica a quantidade de vezes que a regra de estar certa, ou seja, o valor de 80% indica que a regra deve estar certa em 80% das vezes que os parâmetros aparecerem

In [11]:
# Training Apriori on the dataset
from apyori import apriori
rules = apriori(transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2)


In [13]:
# Visualising the results
results = list(rules)

In [14]:
results

[RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'pasta', 'escalope'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

In [17]:
#melhor visualização no SPYDER

results[4]

RelationRecord(items=frozenset({'ground beef', 'herb & pepper'}), support=0.015997866951073192, ordered_statistics=[OrderedStatistic(items_base=frozenset({'herb & pepper'}), items_add=frozenset({'ground beef'}), confidence=0.3234501347708895, lift=3.2919938411349285)])

Pessoa que compra o Groud Beef tem 32% de chance de comprar o herb & pepper